### DISCUSSION TITLE
Alternative to Thonny

### DISCUSSION THREAD

**Discussion by MrTinkerman** [micropython/micropython#12686] [General]
Yeah,

I've used Thonny when getting into uP programming. It served me well, up until now.

I'm looking for an alternative. Thonny just does not handle files, uploads, saves, REPL output that well...
Just too many unknowns, is it my code or is it Thonny being buggy ?? 
In some cases it's my code, in some others it's Thonny f*ing up.

At the moment I'm working on a LoRa project. Two XIOA ESP32-C3 boards with RFM95's hooked up.
I have both ESP's connected to my machine for programming and debugging, Thonny just can't handle that :(
I have the boards communicating over LoRa successfully but when connecting both boards it's just one big mess of unknowns...

So my question is, how do you program your uP devices?

I have tried uPyCraft, it just does not install on my Win10 x64, a missing dll's rabbit hole 
MuEditor installs but cannot find my connected boards...
Pycharm can't find my boards and is actually a no go because of file management.

Is there any stable uP IDE, free or paid, that is reliable?

Would love to hear your feedback,
Thanks

**Comment by davefes** [comment]
Just got VSCode working with the PICO following:
https://randomnerdtutorials.com/raspberry-pi-pico-vs-code-micropython/
Google tells me there is an extension for the ESP32.

I don't know about connecting to multiple boards.

**Comment by davefes** [comment]
Maybe:
https://randomnerdtutorials.com/micropython-esp32-esp8266-vs-code-pymakr/

**Comment by glenn20** [comment]
I find vscode (with @Josverl 's [micropython-esp32-stubs](https://github.com/Josverl/micropython-stubs)) to be a good micropython dev environment for me (and means I have the same environment for other dev). I just use mpremote in a vscode terminal pane to flash and interact with the devices. Like you, I am frequently interacting with multiple devices and I haven't found a vscode plugin that makes that easier than using mpremote (though I haven't looked for a while).

**Reply by fsj-digital** [reply]
`glenn20 

many thanks for sharing  your ideas - well i love VSCode - for its tremendous options to extend. 
Great that there were som manpages and HowTos - that help us to get connected

**Comment by kjm1102** [comment]
I'm still using Atom with a pymakr extension for lora work. It's been abandoned now but I like the way it allows you to have a couple of cards on 2 different com ports & lets you hop back & forth between them as they exchange lora packets with each other.

**Reply by peterhinch** [reply]
You might like to try Pulsar, a fork of Atom which is actively maintained.

**Comment by MrTinkerman** [comment]
Thanks for the suggestions,

I gave mpremote a go in vscode with this plugin: https://marketplace.visualstudio.com/items?itemName=DavesCodeMusings.mpremote
```
\aiolora` Lib\V05> py.exe -m mpremote connect COM4 fs ls /
TransportError:` could not enter raw repl
```
I don't think this is the tool for me, but I'll explore it some more.

I also gave PyMakr a try, I've now got a project with my two devices and two terminal windows.
Looks like what I'm after but figuring out VScode is going to take some time I think.

I found a "workaround" in Thonny though. 
-- Open two Thonny windows, 
-- Start the code on both boards
-- Disconnect one device 
-- Open a system shell window for the other (Tools > Open system shell)
-- Cancel out when it cannot connect
-- The separate system shell now displays incoming serial prints properly

The most workable solution at the moment seems to be hooking the two boards up to two computers.
Luckily I have enough clutter on my big desk to put a laptop on.

**Reply by DavesCodeMusings** [reply]
Hey there! I'm Dave, the guy who wrote the VS Code extension. I've got a similar ESP32-C3 board, I think. The manufacturer is Tourdeus, but I'm pretty sure it's a LOLIN C3 clone. Maybe like yours?

Anyway, I haven't encountered the "transport error" you posted, so I'm not sure what I can offer for help on that. But, one thing I've noticed about the C3 board I have is that there's no USB to Serial chip on it. Looking at a picture of the XIOA, I'm guessing it's the same.

Without the USB-Serial chip, I'm not able to see debug output after resetting the board like I can on my Espressif DevKit C3 board. The Tourdeus simply breaks the connection on reset, where as the Espressif DevKit board keeps the connection and displays boot messages. Maybe you're running into something similar?

**Comment by jimmo** [comment]
@MrTinkerman I would suggest finding a good Python editor/IDE to edit your code (e.g. VSCode is pretty popular and well-supported by the micropython stubs linked above), and use `mpremote` to work with your device. You can even use it from within embedded terminal supported by most IDEs.

I saw your comment here: https://github.com/orgs/micropython/discussions/12657#discussioncomment-7262280
Can you elaborate further -- what can we do to improve the documentation? What else would be helpful?

One of the key problems here is that Damien and I (the two main authors of mpremote) aren't IDE people, so we don't necessarily have a great understanding of what people want from these tools. We built mpremote to be a workflow that works well for us (within the constraints of what we know MicroPython can do), but we are also cognizant of not being the typical users!

I've started some work (but still a long way to go) to try and turn `mpremote` into a useful library/building-block for other people to build more tools (e.g. IDE integration).

**Reply by glenn20** [reply]
Oh, and just to be sure there is no confusion, @jimmo (and I) are referring to the `mpremote` command line tool (docs [here](https://docs.micropython.org/en/latest/reference/mpremote.html)), NOT the `mpremote-vscode` extension by @DavesCodeMusings  for vscode (which uses `mpremote` under the hood). `mpremote` does require you to have python installed on your computer.

From within vscode, you can just open a new `Terminal` from the top menu and install and use `mpremote` there (following the docs linked above). This is what I do.

The `mpremote-vscode` extension may provide a more convenient interface for using `mpremote`.

Whichever IDE you use, if you are having problems connecting to a device, most people here would suggest checking if it works with `mpremote` first (it is the offical tool of micropython).

**Reply by DavesCodeMusings** [reply]
Yep, the mpremote-vscode extension just builds the command-line options for `mpremote` based on what you click, select, etc. and then runs it all in a VS Code terminal window. It's not an official tool or anything. `mpremote` is brilliant in what it does. I just dislike typing command-line options enough that I decided to write an extension as a wrapper around it. I think of it like using Filezilla instead of FTP on the command-line. I also wrote an extension for `esptool` to make firmware flashing a point and click operation. My goal was to make VS Code do everything I had been doing in Thonny, but be able to sync my git repos, something Thonny lacks.

**Reply by MrTinkerman** [reply]
@jimmo @DavesCodeMusings and others, thanks for the replies.

First:
I feel kinda stuck in that twilight zone/gap going from beginner level to somewhat more advanced. I have some coding experience in C++ for Arduino, Python, MicroPyton, Actionscript. But always kinda shied away from the command line. Looking at the  [docs on mpremote](https://docs.micropython.org/en/latest/reference/mpremote.html), it's a bit daunting. I'm not much of a CLI-guy, more like a Select-Options-and-Point-and-Click-Guy. Jimmo, it's not the documentation, it's me not being able to fully understand it.

When reading the docs on mpremote I understand it can be a very useful tool, but.. command line.
DavesCodeMusings has integrated that into VScode, nice work. I'm running into some unknown errors though.
Same goes for PyMakr, integrated into VScode so it would require me to explore and learn a new IDE.

Ok, just brainstorming here! No requests or anything, just posting ideas.

- How about a standalone GUI for mpremote, independent from any IDE?
- Edit your files in your IDE of choice, ctrl-s them locally
- A file browser for local (select directory) and remote files on the device (select device/COMx)
- In the GUI, on every file in the browser, access to all mpremote command via options/buttons

At the moment, Thonny is behaving 'properly' for me, it's workable. 
But I do want to explore the VScode options some more.

Thanks

**Reply by DavesCodeMusings** [reply]
@MrTinkerman I have some sample workflows in the [mpremote-vscode extension wiki](https://github.com/DavesCodeMusings/mpremote-vscode/wiki) that might help you get started. The screenshots alone are worth at least a million words. :-)

I've tried to make the extension intuitive in terms of where you click for what. For example: if you want to enter the REPL prompt, you right-click the COM port that your microcontroller is attached to and select the option from the context menu; if you want to run the micropython code that's in the current editor window, you right-click in the window and select from the MPRemote sub-menu.

The sample workflows will walk through most of this, first using a typical beginner coding scenario of lighting up a NeoPixel, and then a more challenging REST API example that demonstrates installing module dependencies and managing multiple files.

I've been putting my money where my mouth is, so to speak, and using VS Code and my extensions for my projects. However, I only know what works for me, so if you come across something that seems odd, let me know. What seems logical to me may not be so for everyone else. I do try to keep an eye on the [issues list](https://github.com/DavesCodeMusings/mpremote-vscode/issues), but as a dev team of one (in my free time) turnaround will be hit and miss.

Also, please note, my experience is very centered around ESP32 with Windows as the VS Code O.S. I may not be much help if you're using Linux or Mac or some other microcontroller.

**Comment by beetlegigg** [comment]
Inspired by this thread to set up vscode for use with mp I set up a new venv and installed the @DavesCodeMusings mpremote.  That seems to work well.   I then installed the mp stubs into the venv.  I made the settings to the vsc workplace whilst in the venv as per stubs doc (Python language server to Pylance and Pylance Type Checking to strict.
Now I can get, for example, import machine, or from machine import Pin to come up in the list of options when typing the code, but the 'machine' still gets a wiggly line and vscode highlights a problem - "Import "machine" could not be resolved from source"    Does anyone know if this is meant to happed?  Its happy enough to forgo any wiggly lines if I import uasyncio.  
Also, a worrying thought popped into my mind.  Now that to u prefix is dropped, e.g. just import ascynio and not uasyncio for micopython I expect that vscode may get a bit confused twixt python and micropython.   Does anyone have any thoughts whether using vscode for mp is still a good way to proceed?   I was hoping to use it as its my preferred IDE for python with my rpi, and I was thinking It would be good for an mp IDE for the rpi pico's as well. At my level I need all the IDE help I can get :-)

**Reply by glenn20** [reply]
Add

```json
    "python.analysis.diagnosticSeverityOverrides": {
        "reportMissingModuleSource": "none"
    },
```

to your vscode settings.json. If you are relying on stubs for type checking, this says not to worry if the source code doesn't seem to be installed ont he system.

**Reply by glenn20** [reply]
Strict type checking might be a bit challenging for the micropython stubs. I just use "basic" when working with micropython code. @Josverl does a really good job with the micropython stub files. It works great for me - I put type hints on my micropython code. Even though micropython actually ignores them, vscode (and other IDEs) use them to great effect and help me avoid problems.

Although the type hints make the job of the micropython compiler harded and make my python files larger - that all goes away if you compile your python code to `.mpy` files.

**Reply by Josverl** [reply]
My mantra while working on stubber and the stubs has always been : 
**_"Make it not wrong most of the time"_** - which is quite different from **_"Make it right all of the time"_**.
So fully agree: 
- Set typechecks to "Basic" 
- and acknowledge that the type checker cannot find the `.py` en `.c` modules embedded/compiled into your favorite firmware.

Strict type checking may someday be possible - but for now I would like to get the basic coverage up to more than I can conjure out of thin air and partial documentation written for humans rather than for typecheckers.
 
But work on that is underway  by both @jimmo myself and a few others

**Comment by beetlegigg** [comment]
Thanks to @glenn20 & @Josverl for the tips, the vscode is working splendidly and its nice to work with a good IDE for micropython.  Thonny is also good, but on my older mac the latest Thonny is misbehaving where I cannot get to select text by scrolling down to highlight text not currently showing in the window.  Well that maybe a good excuse to get a new computer :)  but vsc is doing very nicely

**Reply by mendenm** [reply]
@beetlegigg On your issue with Thonny on the mac.  Are you using the pip installed Thonny, or the official downloaded, fully wrapped version with its own python.  I was having a terrible time with the pip installed thonny under anaconda.  It was losing about 3/4 of my mouse clicks, and doing all kinds of other strange stuff.  When I downloaded the wrapped package from the Thonny site, it worked perfectly. (Sorry  for posting a comment so long after you posted yours!)

**Reply by beetlegigg** [reply]
@mendenm
I'm using the official download mac dmg file.   My issue is apparently to do with the GUI Toolkit and that a future version of Tk may fix it. I give a link to the issue I raised on the Thonny github.  https://github.com/thonny/thonny/issues/2945

**Comment by JBelmont72** [comment]
I really really greatly greatly appreciate the input I received here. You are all Lifesavers, (the CC syntax issue was just my error in copying it to the GitHub). I have read the links and all of your comments. They’ve all been extremely helpful. I do have VS code set up and think that proceeding on that route may be my best in the long run. I have ordered another couple of Picos to see if my pico is an issue. What was curious is that I had two programs on the pico that I was switching back-and-forth on . They were  Working fine and in the middle of it began the intractable recalcitrant error messages . The pico never recovered even when I used it on another computer.  My next move will be to follow the random nerds tutorial on using the pico on VS code as recommended here. I’m going to try all these different recommendations since I am curious and appreciate the ammunition you have provided in  helping me fight the good fight.

**Comment by vshymanskyy** [comment]
Please check out ViperIDE for Web and Mobile:
https://github.com/orgs/micropython/discussions/15219

Feedback is welcome

**Comment by maruf009sultan** [comment]
mueditor  [https://codewith.mu/en/download] is my fav. {beginner friendly i guess}

**Reply by GitHubsSilverBullet** [reply]
The holy trinity of programming: vi, docs, brain…
